pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go

```diff
@@ -0,0 +1,453 @@
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/hashicorp/go-hclog"
+	"github.com/sigstore/cosign/cmd/cosign/cli/fulcio"
+	"github.com/sigstore/cosign/pkg/cosign"
+	"github.com/sigstore/cosign/pkg/cosign/bundle"
+	"github.com/sigstore/cosign/pkg/oci"
+	rekor "github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/sigstore/pkg/signature/payload"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// Signature Verification Selector
+	signatureVerifiedSelector = "sigstore-validation:passed"
+)
+
+type Sigstore interface {
+	AttestContainerSignatures(ctx context.Context, status *corev1.ContainerStatus) ([]string, error)
+	FetchImageSignatures(ctx context.Context, imageName string) ([]oci.Signature, error)
+	SelectorValuesFromSignature(oci.Signature, string) (*SelectorsFromSignatures, error)
+	ExtractSelectorsFromSignatures(signatures []oci.Signature, containerID string) []SelectorsFromSignatures
+	ShouldSkipImage(imageID string) (bool, error)
+	AddSkippedImage(imageID []string)
+	ClearSkipList()
+	AddAllowedSubject(subject string)
+	EnableAllowSubjectList(bool)
+	ClearAllowedSubjects()
+	SetRekorURL(rekorURL string) error
+	SetLogger(logger hclog.Logger)
+}
+
+// The following structs are used to go through the payload json objects
+type BundleSignature struct {
+	Content   string            `json:"content"`
+	Format    string            `json:"format"`
+	PublicKey map[string]string `json:"publicKey"`
+}
+
+type BundleSpec struct {
+	Data      map[string]map[string]string `json:"data"`
+	Signature BundleSignature              `json:"signature"`
+}
+
+type BundleBody struct {
+	APIVersion string     `json:"apiVersion"`
+	Kind       string     `json:"kind"`
+	Spec       BundleSpec `json:"spec"`
+}
+
+type SelectorsFromSignatures struct {
+	Subject        string
+	Content        string
+	LogID          string
+	IntegratedTime string
+}
+
+func New(cache Cache, logger hclog.Logger) Sigstore {
+	return &sigstoreImpl{
+		functionHooks: sigstoreFunctionHooks{
+			verifyFunction:             cosign.VerifyImageSignatures,
+			fetchImageManifestFunction: remote.Get,
+			checkOptsFunction:          defaultCheckOptsFunction,
+		},
+
+		rekorURL: url.URL{
+			Scheme: rekor.DefaultSchemes[0],
+			Host:   rekor.DefaultHost,
+			Path:   rekor.DefaultBasePath,
+		},
+		logger:        logger,
+		sigstorecache: cache,
+	}
+}
+
+func defaultCheckOptsFunction(rekorURL url.URL) (*cosign.CheckOpts, error) {
+	if rekorURL.Host == "" {
+		return nil, errors.New("rekor URL host is empty")
+	}
+	if rekorURL.Scheme == "" {
+		return nil, errors.New("rekor URL scheme is empty")
+	}
+	if rekorURL.Path == "" {
+		return nil, errors.New("rekor URL path is empty")
+	}
+
+	rootCerts, err := fulcio.GetRoots()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get fulcio root certificates: %w", err)
+	}
+
```

Comment:
I would add `co.IntermediateCerts, err = fulcio.GetIntermediates()`, since the TUF root stores both the roots and intermediates.